table/block_based/block_based_table_builder.cc

```diff
@@ -1569,16 +1598,47 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
         block_rep->first_key_in_next_block->assign(
             r->data_block_and_keys_buffers[i + 1].second.front());
       } else {
-        block_rep->first_key_in_next_block.reset(nullptr);
+        if (r->first_key_in_next_block == nullptr) {
+          block_rep->first_key_in_next_block.reset(nullptr);
+        } else {
+          block_rep->first_key_in_next_block->assign(
+              r->first_key_in_next_block->data(),
+              r->first_key_in_next_block->size());
+        }
       }
 
+      uint64_t new_raw_bytes_inflight =
+          r->pc_rep->raw_bytes_inflight.fetch_add(block_rep->data->size(),
+                                                  std::memory_order_relaxed) +
+          block_rep->data->size();
+      uint64_t new_blocks_inflight =
+          r->pc_rep->blocks_inflight.fetch_add(1, std::memory_order_relaxed) +
+          1;
+      r->pc_rep->estimated_file_size.store(
+          r->get_offset() +
+              static_cast<uint64_t>(
+                  static_cast<double>(new_raw_bytes_inflight) *
+                  r->pc_rep->curr_compression_ratio.load(
+                      std::memory_order_relaxed)) +
+              new_blocks_inflight * kBlockTrailerSize,
+          std::memory_order_relaxed);
```

Comment:
These file size estimation logics and following `first_block` logics are presented in `Flush()` as well.